_posts/2024-03-20-kraft-migration.md

```diff
@@ -0,0 +1,201 @@
+After the update, the brokers need to be rolled one by one to make such configuration changes effective.
+On restart, the brokers register to the KRaft controller quorum and the migration begins.
+The KRaft controller leader copies all metadata from ZooKeeper to the `__cluster_metadata` topic.
+
+![KRaft migration running](/assets/images/posts/2024-03-20-kraft-migration-03-kraft-migration.png)
+
+While the migration is running, you can verify its status by looking at the log on the KRaft controller leader or by checking the `kafka.controller:type=KafkaController,name=ZkMigrationState` metric.
+When the migration is completed, that is the metric value is `MIGRATION`, the brokers are still running in ZooKeeper mode.
+The KRaft controllers are in charge of handling any requests related to metadata changes within the cluster but they keep sending RPCs to the brokers for metadata updates.
+Metadata updates are still copied to ZooKeeper and the cluster is working in a so called "dual-write" mode.
+
+![KRaft dual-write](/assets/images/posts/2024-03-20-kraft-migration-04-kraft-dual-write.png)
+
+#### Restarting the brokers in KRaft mode
+
+The next step is to restart the brokers in KRaft mode and not using ZooKeeper anymore.
+In order to do so, the brokers configuration is updated by removing the connection to ZooKeeper and disabling the migration flag.
+All the brokers are rolled again and, on restart, they are now in full KRaft mode without any connection or usage of ZooKeeper.
+The KRaft controllers are still in "dual-write" mode and any metadata changes are copied to ZooKeeper.
+It is important to highlight this is the last chance to rollback and keep using ZooKeeper to store metadata.
+
+![Brokers only full KRaft](/assets/images/posts/2024-03-20-kraft-migration-05-brokers-kraft.png)
+
+#### Migration finalization
+
+The final step is to reconfigure the KRaft controllers without the connection to ZooKeeper and disabling the migration flag.
+When all the KRaft controllers are rolled, the cluster is working in full KRaft mode and the ZooKeeper ensemble is not used anymore.
+From now on, you should deprovision the ZooKeeper nodes from your environment.
+
+![Cluster full KRaft](/assets/images/posts/2024-03-20-kraft-migration-06-kraft-cluster.png)
+
+#### Rollback support
+
+During the migration process is also possible to execute a rollback procedure to revert the cluster to use the ZooKeeper ensemble again.
+The rollback is allowed until the KRaft controllers are still connected to ZooKeeper because they are working in "dual-write" mode by effectively copying metadata to ZooKeeper.
+When the migration is finalized and the KRaft controllers are not connected to ZooKeeper, the rollback is not possible anymore.
+
+### Conclusion
+
+The Apache Kafka community has deprecated the usage of ZooKeeper to store the cluster metadata and it will be totally removed in the 4.0 version coming later this year or early the next one.
```

Comment:
```suggestion
The Apache Kafka community has deprecated the usage of ZooKeeper to store the cluster metadata and it will be completely removed in the 4.0 version coming later this year or early the next one.
```